src/extensions/GetStackTrace.ts

```diff
@@ -0,0 +1,66 @@
+import {
+    BrsString,
+    RoArray,
+    Callable,
+    ValueKind,
+    StdlibArgument,
+    Int32,
+    BrsInvalid,
+    BrsType,
+    isBrsString,
+} from "../brsTypes";
+import type { Location } from "../lexer";
+import { Interpreter } from "../interpreter";
+
+const INTERNAL_REGEX_FILTER = /\(internal\)/;
+
+function stringifyLocation(location: Location) {
+    return `${location.file}:${location.start.line}:${location.start.column}`;
+}
+
+/**
+ * Returns a stack trace in the format:
+ * [
+ *   filename:line:column,
+ *   ...
+ * ]
+ */
+export const GetStackTrace = new Callable("GetStackTrace", {
+    signature: {
+        args: [
+            new StdlibArgument("numEntries", ValueKind.Int32, new Int32(10)),
+            new StdlibArgument("excludePatterns", ValueKind.Dynamic, BrsInvalid.Instance),
+        ],
+        returns: ValueKind.Object,
+    },
+    impl: (interpreter: Interpreter, numEntries: Int32, excludePatterns: BrsType) => {
+        let stack = interpreter.stack;
+
+        // Filter out any files that the consumer doesn't want to include
+        if (excludePatterns instanceof RoArray) {
+            excludePatterns.getValue().forEach((pattern) => {
+                if (isBrsString(pattern)) {
+                    let regexFilter = new RegExp(pattern.value);
+                    stack = stack.filter((location) => !regexFilter.test(location.file));
+                }
+            });
+        } else if (!(excludePatterns instanceof BrsInvalid)) {
+            interpreter.stderr.write("Patterns to exclude argument must be an roArray");
+        }
+
+        return new RoArray(
+            stack
+                // Filter out any internal stack traces.
+                .filter((location) => !INTERNAL_REGEX_FILTER.test(location.file))
+                // Remove any duplicate entries that appear next to each other in the stack.
+                .filter((location, index, locations) => {
+                    if (index === 0) return true;
+                    let prevLocation = locations[index - 1];
+                    return stringifyLocation(prevLocation) !== stringifyLocation(location);
+                })
+                .map((location) => new BrsString(stringifyLocation(location)))
```

Comment:
Watch out!  It looks like this'll stringify every location three times (except for the first and last ones) — once as `stringifyLocation(location)`, again in `new BrsString(stringifyLocation(location))` and once as the "previous location".  If formatting all locations to a string is necessary to remove duplicate entries, consider reordering this so that we stringify everything in one pass then filter the already-converted elements:

```typescript
stack
     // Filter out any internal stack traces.
     .filter((location) => !INTERNAL_REGEX_FILTER.test(location.file))
      .map((location) => new BrsString(stringifyLocation(location)))
     // Remove any duplicate entries that appear next to each other in the stack.
     .filter((location, index, locations) => {
         if (index === 0) return true;
         let prevLocation = locations[index - 1];
         return prevLocation !== location;
      });
```

Alternatively, I suppose you could implement a comparison function on `Location` and call that before converting to strings?  Something like:

```typescript
class Location {
    equalTo(other: Location): boolean {
        return this.file === other.file && this.start.line === other.start.line && this.start.column === other.start.column && …
    }
}
```

But it looks like we're only concerned with the starting locations here anyway, so perhaps that's overkill.

Reply:
Great callout! Yeah, I considered implementing a comparison function on `Location` as well -- that might be the cleanest way to do it.